### PR TITLE
PLT-6324: Uppercase letter required for a password if password requirement set to at least 5 characters and a number

### PR DESCRIPTION
**Pull request by MusikPolice**
#### Summary
The regex for uppercase/numbers were reversed. Fixed the problem, added unit tests so it doesn't happen again in the future.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6324

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: utils/utils.jsx [hunk 1]
```diff
@@ -1333,50 +1333,63 @@ export function canCreateCustomEmoji(user) {
...
+        isEnterprise: global.window.mm_config.BuildEnterpriseReady === 'true',
+        isLicensed: global.window.mm_license.IsLicensed === 'true',
+        isPasswordRequirements: global.window.mm_license.PasswordRequirements === 'true',
+        minimumLength: parseInt(global.window.mm_config.PasswordMinimumLength, 10),
+        requireLowercase: global.window.mm_config.PasswordRequireLowercase === 'true',
+        requireUppercase: global.window.mm_config.PasswordRequireUppercase === 'true',
+        requireNumber: global.window.mm_config.PasswordRequireNumber === 'true',
+        requireSymbol: global.window.mm_config.PasswordRequireSymbol === 'true',
+    }
+}
```
Comment (MusikPolice) [inline utils/utils.jsx]: By hiding the static access in this helper function, my test suites can easily sub in config, and the `isValidPassword(...)` all type coercion can be factored out of `isValidPassword(...)`, which makes it easier to understand.

File: utils/utils.jsx [hunk 2]
```diff
@@ -1333,50 +1333,63 @@ export function canCreateCustomEmoji(user) {
...
                 error = true;
             }
 
             errorId += 'Lowercase';
         }
 
-        if (global.window.mm_config.PasswordRequireUppercase === 'true') {
-            if (!password.match(/[0-9]/)) {
+        if (passwordConfig.requireUppercase) {
+            if (!password.match(/[A-Z]/)) {
```
Comment (MusikPolice) [inline utils/utils.jsx]: This was the actual bug. Everything else is testing

File: tests/utils/utils.test.jsx [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+                    isEnterprise: false,
+                    isLicensed: true,
+                    isPasswordRequirements: true,
+                },
+                valid: true,
+            },
+            {
+                password: 'four',
+                config: { // not EE, so password just has to be min < length < max
+                    isEnterprise: true,
```
Comment (saturninoabril) [inline tests/utils/utils.test.jsx]: Should this be EE?

Comment (MusikPolice) [reply tests/utils/utils.test.jsx]: Yes. This suite turns one of the three booleans off for each set of three cases. So in the first three tests, `isEnterprise` is false, but in this set of three tests, `isLicensed` is false. In the last set of three tests, `isPasswordRequirements` is false. All three flags must be true for the EE passwords requirements to be checked, so this is the only way to check that the requirements are bypassed in all cases that they should be.

Comment (saturninoabril) [reply tests/utils/utils.test.jsx]: Ah okay. I mean, the comment `not EE` might need to remove?

**Comment by MusikPolice** [comment]
Sorry about that. Fixed

**Pull request merged**